rasa/core/policies/ted_policy.py

```diff
@@ -381,21 +567,72 @@ def predict_action_probabilities(
             return self._prediction(self._default_predictions(domain))
 
         # create model data from tracker
-        tracker_state_features = self.featurizer.create_state_features(
-            [tracker], domain, interpreter
+        tracker_state_features = self._featurize_tracker_for_e2e(
+            tracker, domain, interpreter
         )
         model_data = self._create_model_data(tracker_state_features)
 
         output = self.model.predict(model_data)
 
-        confidence = output["action_scores"].numpy()
-        # remove batch dimension and take the last prediction in the sequence
-        confidence = confidence[0, -1, :]
+        # take the last prediction in the sequence
+        similarities = output["similarities"].numpy()[:, -1, :]
+        confidences = output["action_scores"].numpy()[:, -1, :]
+        # take correct prediction from batch
+        confidence, is_e2e_prediction = self._pick_confidence(confidences, similarities)
 
         if self.config[LOSS_TYPE] == SOFTMAX and self.config[RANKING_LENGTH] > 0:
             confidence = train_utils.normalize(confidence, self.config[RANKING_LENGTH])
 
-        return self._prediction(confidence.tolist())
+        optional_events = self._create_optional_event_for_entities(
+            output, interpreter, tracker
+        )
+
+        return self._prediction(
+            confidence.tolist(),
+            is_end_to_end_prediction=is_e2e_prediction,
+            optional_events=optional_events,
+        )
+
+    def _create_optional_event_for_entities(
+        self,
+        prediction_output: Dict[Text, tf.Tensor],
+        interpreter: NaturalLanguageInterpreter,
+        tracker: DialogueStateTracker,
+    ) -> Optional[List[Event]]:
+        if tracker.latest_action_name != ACTION_LISTEN_NAME:
+            # entities belong to the last user message
+            # a user message is always followed by action listen
```

Comment:
I guess you mean vice versa
```suggestion
            # a user message always comes after action listen
```